src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+                
+                public MyList(params int[] values) : base()
+                {
+                    Values = values;
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list1 = [with(), 1];
+                    MyList<int> list2 = [with(10), 2];
+                    MyList<int> list3 = [with(10, 20, 30), 3];
+                    
+                    Console.WriteLine($"{list1.Values.Length},{list2.Values.Length},{list3.Values.Length}");
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("0,1,3"));
+    }
+
+    [Fact]
+    public void WithElement_ParamsWithNamedArguments()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public string Name { get; }
+                public int[] Values { get; }
+                
+                public MyList(string name, params int[] values) : base()
+                {
+                    Name = name;
+                    Values = values;
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(name: "test", values: new int[] { 1, 2, 3 }), 4];
+                    Console.WriteLine($"{list.Name},{list.Values.Length}");
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("test,3"));
+    }
+
+    [Fact]
+    public void WithElement_ParamsCollection()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            using System.Linq;
+            
+            class MyList<T> : List<T>
+            {
+                public IEnumerable<int> Values { get; }
+                
+                public MyList(params IEnumerable<int> values) : base()
+                {
+                    Values = values;
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(new int[] { 1, 2, 3 }), 4];
+                    Console.WriteLine($"{list.Values.Count()}");
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("3"));
+    }
+
+    #endregion
+
+    #region Dynamic Tests
+
+    [Fact]
+    public void WithElement_DynamicArguments()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public object Value { get; }
+                
+                public MyList(object value) : base()
```

Comment:
Consider testing when the argument type is actually `dynamic` as well, passing both `dynamic` and non-`dynamic` arguments.

Reply:
e7bc129ad3f